packages/go_router/lib/src/builder.dart

```diff
@@ -413,7 +424,14 @@ class _CustomNavigatorState extends State<_CustomNavigator> {
   bool _handlePopPage(Route<Object?> route, Object? result) {
     final Page<Object?> page = route.settings as Page<Object?>;
     final RouteMatchBase match = _pageToRouteMatchBase[page]!;
-    return widget.onPopPageWithRouteMatch(route, result, match);
+    final bool didPop = widget.onPopPageWithRouteMatch(route, result, match);
+    if (didPop) {
+      _pages?.remove(page);
```

Comment:
seems slightly dangerous we do it this way. One way to avoid this is that `_CustomNavigatorState` listens to GoRouterDelegate, which `widget.onPopPageWithRouteMatch` should cause the GoRouterDelegate to notify its listener. In the listener callback, we reconstruct `_pages` and `_pageToRouteMatchBase` as well as `_registry`.

If we do that, we have to access routematchlist directly from goRouterDelegate

Reply:
I meant we may have a temporary difference between the states and route match. it may be hard to keep them in sync.

Reply:
> If we do that, we have to access routematchlist directly from goRouterDelegate

Actually, that is a bit harder than that because of nested shell routes. What we really need is the `List<RouteMatchBase> matches` to construct the pages.
However, right now, this is passed as a parameter for `_CustomNavigator`:

https://github.com/flutter/packages/blob/6a4b660482cbd2a04e55c26ccff61f404b1cbbc1/packages/go_router/lib/src/builder.dart#L128-L139

Adding a listener to the `GoRouterDelegate` would notify any changes. With the current implementation, there is no way to know if the changes are supposed to be reflected to the current `_CustomNavigator`'s pages. And if yes, how nested we are  in the `RouteMatchList`: the `_CustomNavigator` is not necessarily the root navigator. So we cannot use `widget.matchList.matches`

Or maybe we can have a chat on discord or a call to talk how you are seeing it?

Reply:
The `required this.matches` probably should be removed, since we will need to get the latest matches in the State.

I think we may be able to have a way to look up matches from the matchList by using navigatorKey, or maybe a the routerdelegate can pass a look up callback to `_CustomNavigator` instead of of a fixed List<RouteMatches>?